nowcasting_dataset/data_sources/data_source.py

```diff
@@ -112,13 +121,90 @@ def open(self):
+        pass
+
+    def check_input_paths_exist(self) -> None:
+        """Check any input paths exist.  Raise FileNotFoundError if not.
 
-        Data sources which can be forked safely should call open()
-        from __init__().
+        Can be overridden by child classes.
         """
         pass
 
+    def create_batches(
+        self,
+        spatial_and_temporal_locations_of_each_example: pd.DataFrame,
+        idx_of_first_batch: int,
+        batch_size: int,
+        dst_path: Path,
+        local_temp_path: Path,
+        upload_every_n_batches: int,
+    ) -> None:
+        """Create multiple batches and save them to disk.
+
+        Safe to call from worker processes.
+
+        Args:
+          spatial_and_temporal_locations_of_each_example: A DataFrame where each row specifies
+            the spatial and temporal location of an example.  The number of rows must be
+            an exact multiple of `batch_size`.
+            Columns are: t0_datetime_UTC, x_center_OSGB, y_center_OSGB.
+          idx_of_first_batch: The batch number of the first batch to create.
+          batch_size: The number of examples per batch.
+          dst_path: The final destination path for the batches.  Must exist.
+          local_temp_path: The local temporary path.  This is only required when dst_path is a
+            cloud storage bucket, so files must first be created on the VM's local disk in temp_path
+            and then uploaded to dst_path every upload_every_n_batches. Must exist. Will be emptied.
+          upload_every_n_batches: Upload the contents of temp_path to dst_path after this number
+            of batches have been created.  If 0 then will write directly to dst_path.
+        """
+        # Sanity checks:
+        assert idx_of_first_batch >= 0
+        assert batch_size > 0
+        assert len(spatial_and_temporal_locations_of_each_example) % batch_size == 0
+        assert upload_every_n_batches >= 0
+
+        self.open()
+
+        # Figure out where to write batches to:
+        save_batches_locally_and_upload = upload_every_n_batches > 0
+        if save_batches_locally_and_upload:
+            nd_fs_utils.delete_all_files_in_temp_path(local_temp_path)
+        path_to_write_to = local_temp_path if save_batches_locally_and_upload else dst_path
+
+        # Loop round each batch:
+        examples_for_batch = spatial_and_temporal_locations_of_each_example.iloc[:batch_size]
+        n_batches_processed = 0
+        while not examples_for_batch.empty:
+            # Generate batch.
+            batch = self.get_batch(
+                t0_datetimes=examples_for_batch.t0_datetime_UTC,
+                x_locations=examples_for_batch.x_center_OSGB,
+                y_locations=examples_for_batch.y_center_OSGB,
+            )
+
+            # Save batch to disk.
+            batch_idx = idx_of_first_batch + n_batches_processed
+            netcdf_filename = path_to_write_to / nd_utils.get_netcdf_filename(batch_idx)
```

Comment:
worth using os.path.join?

Reply:
im always funny about / and \\ and paths that have / at the end .e.t.c

Reply:
Ah, so, the nice thing about `Path` objects is that `Path` should always use the correct directory separator (`/` or `\`) depending on the platform.  `Path` overwrites the `/` operator to behave like `os.path.join`.

I'm not entirely certain but I think `pathlib.Path` is meant to be a modern replacement for `os.path`.